src/transport/MessageHeader.cpp

```diff
@@ -158,60 +165,63 @@ CHIP_ERROR PacketHeader::Decode(const uint8_t * data, size_t size, size_t * deco
 
     if (mFlags.value & Header::Flags::kDestinationNodeIdPresent)
     {
-        VerifyOrExit(size >= kNodeIdSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrExit(size >= (p - data) + kNodeIdSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
         mDestinationNodeId.SetValue(LittleEndian::Read64(p));
-        size -= kNodeIdSizeBytes;
     }
     else
     {
         mDestinationNodeId.ClearValue();
     }
 
-    VerifyOrExit(size >= sizeof(uint16_t) * 2, err = CHIP_ERROR_INVALID_ARGUMENT);
+    static_assert(CHAR_BIT == 8, "We're assuming sizeof(uint16_t) is 2 bytes");
+    VerifyOrExit(size >= (p - data) + sizeof(uint16_t) * 2, err = CHIP_ERROR_INVALID_ARGUMENT);
     mEncryptionKeyID = LittleEndian::Read16(p);
     mPayloadLength   = LittleEndian::Read16(p);
 
-    *decode_len = p - data;
+    VerifyOrExit(p - data == EncodeSizeBytes(), err = CHIP_ERROR_INTERNAL);
+    static_assert(std::is_same<decltype(EncodeSizeBytes()), uint16_t>::value, "Unexpected EncodeSizeBytes() return type");
+    *decode_len = static_cast<uint16_t>(p - data);
 
 exit:
 
     return err;
 }
 
-CHIP_ERROR PayloadHeader::Decode(Header::Flags flags, const uint8_t * data, size_t size, size_t * decode_len)
+CHIP_ERROR PayloadHeader::Decode(Header::Flags flags, const uint8_t * const data, size_t size, uint16_t * decode_len)
 {
     CHIP_ERROR err    = CHIP_NO_ERROR;
     const uint8_t * p = data;
 
-    // TODO: this seems odd - know the size before decoding
-    // likely because header cross-referencing
-    VerifyOrExit(size >= EncodeSizeBytes(), err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(size >= kEncryptedHeaderSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
 
     mExchangeHeader = Read8(p);
     mMessageType    = Read8(p);
     mExchangeID     = LittleEndian::Read16(p);
 
     if (flags.value & Header::Flags::kVendorIdPresent)
     {
+        VerifyOrExit(size >= (p - data) + kVendorIdSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
         mVendorId.SetValue(LittleEndian::Read16(p));
     }
     else
     {
         mVendorId.ClearValue();
     }
 
+    static_assert(CHAR_BIT == 8, "We're assuming sizeof(uint16_t) is 2 bytes");
```

Comment:
This is unclear: why is CHAR_BIT used for read16? we should compare sizeof(uint16_t) to ensure read16 is ok. Or even skip since uint16_t not being 16 bit or read16 not reading exactly 2 bytes would be a fundamental error.

Reply:
`CHAR_BIT ` is not used for `read16`.  But it's used for `sizeof`!

That is, `read16` will always read 16 bits, and `uint16_t` is 16 bits.  But `uint16_t` being "2 bytes" in the C/C++ definition of "1 byte is 1 `char`" is _not_ guaranteed, in general.  For example, you could have a platform with 16-bit `char`; then `uint16_t` will be 1 byte and `sizeof(uint16_t) == 1` will be true.

CHIP does depend on `CHAR_BIT == 8` in various places, but I feel that we should explicitly assert it in the places where we depend on it, so that in case someone decides to try to port CHIP to a platform where this does not hold they will at least know where they need to figure out how to change stuff instead of running into silent bugs.

Reply:
I would move the static_assert on CHAR_BIT _sxomewwhere else, as it does not seem related to uint16_t.
We can replace if needed with a check sizeof(uint16_t) == 2, which will relate with the following line.

Reply:
Per offline discussion, will just remove this static assert.  I expect that the issue will become moot when we fix https://github.com/project-chip/connectedhomeip/issues/2914